docs: Add note about computation perturbation in debug.print/breakpoint

alhridoy opened the pull request:
This commit adds documentation about how using jax.debug.print and jax.debug.breakpoint can perturb the computation that XLA compiles, potentially leading to numeric discrepancies.

Fixes #26370

dfm commented:
Thanks! Can you fill out the CLA as requested by the bot?

alhridoy commented:
@dfm Thank you! I signed.

dfm commented:
Thanks! Although, it looks like there might be a mismatch between the email address you have configured for git, and the one you used to sign the form. Can you click through to that failing workflow for more info, and try to get it green?

alhridoy commented:
> Thanks! Although, it looks like there might be a mismatch between the email address you have configured for git, and the one you used to sign the form. Can you click through to that failing workflow for more info, and try to get it green?

Thank you! I think it is ok now.

dfm commented on docs/debugging/print_breakpoint.md:
```diff
@@ -177,6 +177,10 @@ Why? Under the hood, the compiler gets a functional representation of the staged
 
+#### Computation perturbation
+
+Adding `jax.debug.print` or `jax.debug.breakpoint` statements will change the computation that XLA is asked to compile. This can potentially result in numeric discrepancies compared to the same code without debug statements. Keep this in mind when debugging numerical issues, as the act of adding debug statements might affect the behavior you're trying to investigate.
```
It might be worth explicitly noting that the reason for this change is that XLA might perform different fusions when compiling.

dfm commented:
@alhridoy — Sorry, I didn't see that you had pushed a change because I only get notified for comments. Looks good now. Can you [squash and rebase your commits onto the current main branch](https://docs.jax.dev/en/latest/contributing.html#single-change-commits-and-pull-requests)? Once you've done that, please ping me and I'll get this merged. Thanks!!

alhridoy committed:
docs: Add note about computation perturbation in debug.print/breakpoint

This commit adds documentation about how using jax.debug.print and
jax.debug.breakpoint can perturb the computation that XLA compiles,
potentially leading to numeric discrepancies. This happens because XLA
might perform different fusions when compiling with debug statements.

Fixes #26370

superbobry reviewed (approved):
Thanks!
